[SPARK-23011][SQL][PYTHON] Support alternative function form with group aggregate pandas UDF

icexelloss opened the pull request:
## What changes were proposed in this pull request?

This PR proposes to support an alternative function from with group aggregate pandas UDF.

The current form:
```
def foo(pdf):
    return ...
```
Takes a single arg that is a pandas DataFrame.

With this PR, an alternative form is supported:
```
def foo(key, pdf):
    return ...
```
The alternative form takes two argument - a tuple that presents the grouping key, and a pandas DataFrame represents the data.

## How was this patch tested?

GroupbyApplyTests

icexelloss committed:
Initial commit. Test passes.

SparkQA commented:
**[Test build #86286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86286/testReport)** for PR 20295 at commit [`38195ac`](https://github.com/apache/spark/commit/38195ac7f0b9e1227cfc1e407de47e276b3fc43f).

SparkQA commented:
**[Test build #86286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86286/testReport)** for PR 20295 at commit [`38195ac`](https://github.com/apache/spark/commit/38195ac7f0b9e1227cfc1e407de47e276b3fc43f).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86286/
Test FAILed.

icexelloss committed:
Add comments explaining arg offsets

icexelloss committed:
Fix timestamp conversion

SparkQA commented:
**[Test build #86290 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86290/testReport)** for PR 20295 at commit [`7ce3fa7`](https://github.com/apache/spark/commit/7ce3fa79f8f63d4320c386ed81fa0b57d6bb7a91).

icexelloss commented:
cc @ueshin @HyukjinKwon @cloud-fan @viirya 

This PR implements discussion here https://github.com/apache/spark/pull/20211#pullrequestreview-87657832. There are more refinement needs to be done but I'd like to get some early feedback whether this approach looks good  in general.

The general idea is to pass grouping columns as extra columns to the python worker and to use `argOffsets` to specify which columns are grouping columns. Finally, we convert a grouping columns to a single tuple before enter user function. This is slightly inefficient because grouping columns always have the same value for each group. But I think this is OK because grouping columns should be relatively small comparing to the entire DataFrame.

I can also implement some kind of de duplicate logic in `FlatMapGroupsInPandasExec`, however that would require creating another `UnsafeProjection`, I am not sure if it's worth it performance wise. 

WDYT?

SparkQA commented:
**[Test build #86290 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86290/testReport)** for PR 20295 at commit [`7ce3fa7`](https://github.com/apache/spark/commit/7ce3fa79f8f63d4320c386ed81fa0b57d6bb7a91).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86290/
Test FAILed.

cloud-fan commented:
How are you going to send the group columns? For a group we have only one group row and a bunch of data rows.

icexelloss commented:
@cloud-fan Currently I sent group columns along with the extra data column. For example, if the original DataFrame has `id, v` and group column is `id`, the current implementation in this PR will send three series `id, id, v` to the python worker and send an `argOffsets` of `[1, 2]` to specify the data columns are `id, v`. The first value of the group column is used as the group key, because values in a group column are equal.

I implemented it this way because it doesn't change the existing serialization protocol. Alternatively, we can implement a new serialization protocol for GROUP_MAP eval type, i.e, instead of sending an arrow batch, we could send a group row and then an arrow batch. What do you think?

cloud-fan commented:
How do we turn a single group column to a series? just repeat the group column?

icexelloss commented:
Yep, that's correct.

HyukjinKwon commented:
To me, seems roughly fine.

> Alternatively, we can implement a new serialization protocol for GROUP_MAP eval type, i.e, instead of sending an arrow batch, we could send a group row and then an arrow batch.

I don't have a strong preference on this.

ueshin commented on python/pyspark/serializers.py:
```diff
@@ -267,13 +267,13 @@ def load_stream(self, stream):
-            pdf = _check_dataframe_localize_timestamps(batch.to_pandas(), self._timezone)
-            yield [c for _, c in pdf.iteritems()]
+            yield [_check_series_localize_timestamps(c.to_pandas(), self._timezone)
+                   for c in pa.Table.from_batches([batch]).itercolumns()]
```
Maybe we can remove the comment above (`# NOTE: ...`) ?

icexelloss replied:
I actually don't know what the comment above means. @BryanCutler do you remember?

BryanCutler replied:
Yeah, the note was because prev we iterated over Arrow columns and converted each to a Series, then changed to convert an Arrow batch to DataFrame and then iterated over DataFrame columns to get a Series.  I wasn't sure if there might be a perf decrease, so I left the note but I'm not sure why it wasn't done like the above in the first place - seems like it would be just as good as the original.  Anyway, yeah the note can be removed now.

icexelloss replied:
@BryanCutler Thanks for the clarification. I removed the note.

ueshin reviewed (commented):
I'm also neutral basically, but I'd prefer the new serialization if there is a simple way and performant enough.

icexelloss commented:
Let me experiment with new serialization approach. Will update here.

BryanCutler reviewed (commented):
This sounds fine with me, I think that different serialization is slightly better to avoid duplicate data. Would the group `Row` be sent as a separate Arrow batch?

Regarding the API, I missed the original discussion, but just as an additional thought, a while back I proposed having an optional `kwargs` to each pandas_udf to deal with 0-param udfs. If we were to do that, the group Row could be placed in there and then there wouldn't need to be 2 types of signatures to allow for an optional `key` arg.  I can see why it might be preferable to have an explicit `key` though, so it's up to you guys - just thought I'd mention this again.

icexelloss committed:
Implement deduplicate logic; Add tets

icexelloss committed:
Add docs

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/203/
Test PASSed.

SparkQA commented:
**[Test build #86604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86604/testReport)** for PR 20295 at commit [`c7fccde`](https://github.com/apache/spark/commit/c7fccde701723c8808b095586e987ad02fc171c7).

icexelloss committed:
Remove old notes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/205/
Test PASSed.

SparkQA commented:
**[Test build #86606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86606/testReport)** for PR 20295 at commit [`cda97f1`](https://github.com/apache/spark/commit/cda97f142285bc01a2da96765d3703d37b4cb802).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/206/
Test PASSed.

icexelloss commented:
Hi all,

I did some digging and I think adding a serialization form that serialize a key object along with a Arrow record batch is quite complicated because we are using ArrowStreamReader/Writer for sending batches and send extra key data would have to use a lower level Arrow API for sending/receiving batches.

I did two things to convince myself the current approach is fine:
* I add logic to de duplicate grouping key they are already in data columns. i.e., if a user calls
```
df.groupby('id').apply(foo_udf)
```
We will not send extra grouping columns because those are already part of data columns. Instead, we will just use the corresponding data column to get grouping key to pass to user function. However, if user calls:
```
df.groupby(df.id % 2).apply(foo_udf)
```
then an extra column `df.id % 2` will be created and sent to python worker. But I think this is an uncommon case.

* I did some benchmark to see the impact of sending extra grouping column. I used a Spark DataFrame of a single column to maximize the effect of the extra grouping column (basically sending extra grouping column will double the data to be sent to python in the benchmark, however in real use cases the effect of sending extra grouping columns should be far less).
Even with the setting of the benchmark, I have not observed performance diffs when sending extra grouping columns, I think this is because the total time is dominated by other parts of the computation. [micro benchmark](https://gist.github.com/icexelloss/88f6c6fdaf04aac39d68d74cd0942c07)

I'd like to leave the work for more flexible arrow serialization as future work because it doesn't seems to affect performance of this patch and proceed with the current patch based on the two points above. What do you guys think?

SparkQA commented:
**[Test build #86606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86606/testReport)** for PR 20295 at commit [`cda97f1`](https://github.com/apache/spark/commit/cda97f142285bc01a2da96765d3703d37b4cb802).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86606/
Test FAILed.

SparkQA commented:
**[Test build #86604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86604/testReport)** for PR 20295 at commit [`c7fccde`](https://github.com/apache/spark/commit/c7fccde701723c8808b095586e987ad02fc171c7).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86604/
Test FAILed.

icexelloss committed:
Fix tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/241/
Test PASSed.

SparkQA commented:
**[Test build #86651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86651/testReport)** for PR 20295 at commit [`6b8fdbc`](https://github.com/apache/spark/commit/6b8fdbc61ce8ef03a5ac54d65babc29c0697c9c4).

SparkQA commented:
**[Test build #86651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86651/testReport)** for PR 20295 at commit [`6b8fdbc`](https://github.com/apache/spark/commit/6b8fdbc61ce8ef03a5ac54d65babc29c0697c9c4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86651/
Test PASSed.

HyukjinKwon commented:
For https://github.com/apache/spark/pull/20295#issuecomment-360297123, I am fine without new serialization protocol actually. I didn't have a strong preference there because I wasn't sure if it's worth - complexity vs actual gain vaguely and seems that's now clarified there. I am okay with the current approach.

@BryanCutler, I think the intention here is to follow other few APIs and `gapply` in R. I guess you meant the length and metadata stuff by "an optional kwargs to each pandas_udf to deal with 0-param udfs" if I remember correctly. I think that's slightly different because here the motivation is to provide consistent support similarly with other APIs vs the `kwargs` thing sounds pretty few concept to me.

icexelloss commented:
@HyukjinKwon Thanks for the comment. I will continue with the current approach unless objection raises. I will work on comments and refinements in the next day or two.

icexelloss committed:
Update error message for group map pandas UDF

icexelloss commented on python/pyspark/sql/udf.py:
```diff
@@ -54,7 +54,7 @@ def _create_udf(f, returnType, evalType):
+        if evalType == PythonEvalType.SQL_PANDAS_GROUP_MAP_UDF and len(argspec.args) not in (1, 2):
             raise ValueError(
                 "Invalid function: pandas_udfs with function type GROUP_MAP "
                 "must take a single arg that is a pandas DataFrame."
```
Aha good catch. Fixed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/340/
Test PASSed.

SparkQA commented:
**[Test build #86773 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86773/testReport)** for PR 20295 at commit [`edb77dc`](https://github.com/apache/spark/commit/edb77dc06ba0c675ab2001548d620d5938cd26c0).

icexelloss commented:
@HyukjinKwon @ueshin This is ready for review. I addressed the comments so far.

@BryanCutler yeah I think kwargs is another option. But I think the API in this PR is more consistent with the exsiting APIs though.

SparkQA commented:
**[Test build #86773 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86773/testReport)** for PR 20295 at commit [`edb77dc`](https://github.com/apache/spark/commit/edb77dc06ba0c675ab2001548d620d5938cd26c0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86773/
Test PASSed.

icexelloss committed:
Fix merge error

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/393/
Test PASSed.

SparkQA commented:
**[Test build #86834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86834/testReport)** for PR 20295 at commit [`2668251`](https://github.com/apache/spark/commit/266825167f0bf308c0b4213b1ef718a930a47c2b).

icexelloss commented:
Rebased

SparkQA commented:
**[Test build #86834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86834/testReport)** for PR 20295 at commit [`2668251`](https://github.com/apache/spark/commit/266825167f0bf308c0b4213b1ef718a930a47c2b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86834/
Test FAILed.

icexelloss committed:
Fix style

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/395/
Test PASSed.

SparkQA commented:
**[Test build #86836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86836/testReport)** for PR 20295 at commit [`2399b77`](https://github.com/apache/spark/commit/2399b770551bcc16721af0199971b5b66536707b).

SparkQA commented:
**[Test build #86836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86836/testReport)** for PR 20295 at commit [`2399b77`](https://github.com/apache/spark/commit/2399b770551bcc16721af0199971b5b66536707b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86836/
Test FAILed.

icexelloss committed:
Fix style

SparkQA commented:
**[Test build #86837 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86837/testReport)** for PR 20295 at commit [`8f0782c`](https://github.com/apache/spark/commit/8f0782c07f4c6f02610918e6d4edc5907f7d6aaa).

SparkQA commented:
**[Test build #86837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86837/testReport)** for PR 20295 at commit [`8f0782c`](https://github.com/apache/spark/commit/8f0782c07f4c6f02610918e6d4edc5907f7d6aaa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86837/
Test PASSed.

BryanCutler commented:
> @BryanCutler yeah I think kwargs is another option. But I think the API in this PR is more consistent with the exsiting APIs though.

Yeah if it's consistent with other APIs then sounds fine with me.  My concern was in giving the user too many options that it starts to get confusing to make UDFs. If it's a familiar API then that probably won't be the case.

icexelloss commented:
Hey @HyukjinKwon @ueshin could you please take a look at this? Thanks!

icexelloss committed:
Fix tests

icexelloss commented:
gentle ping @HyukjinKwon @ueshin

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/641/
Test PASSed.

SparkQA commented:
**[Test build #87129 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87129/testReport)** for PR 20295 at commit [`fef4989`](https://github.com/apache/spark/commit/fef49892f6c52c0d6313d10e8d030dc8f85a636b).

SparkQA commented:
**[Test build #87129 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87129/testReport)** for PR 20295 at commit [`fef4989`](https://github.com/apache/spark/commit/fef49892f6c52c0d6313d10e8d030dc8f85a636b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87129/
Test PASSed.

ueshin commented on python/pyspark/sql/types.py:
```diff
@@ -1694,6 +1694,13 @@ def from_arrow_schema(arrow_schema):
          for field in arrow_schema])
 
 
+def _check_series_convert_date(series, data_type):
```
Do we need some documents here as well?

ueshin reviewed (commented):
LGTM so far. Could you fix the conflicts please?

icexelloss committed:
Add docstring

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/900/
Test PASSed.

SparkQA commented:
**[Test build #87455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87455/testReport)** for PR 20295 at commit [`cf45827`](https://github.com/apache/spark/commit/cf458279c4630ed0fb565b6836d10b909185309e).

SparkQA commented:
**[Test build #87455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87455/testReport)** for PR 20295 at commit [`cf45827`](https://github.com/apache/spark/commit/cf458279c4630ed0fb565b6836d10b909185309e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87455/
Test FAILed.

SparkQA commented:
**[Test build #87463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87463/testReport)** for PR 20295 at commit [`cf45827`](https://github.com/apache/spark/commit/cf458279c4630ed0fb565b6836d10b909185309e).

SparkQA commented:
**[Test build #87463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87463/testReport)** for PR 20295 at commit [`cf45827`](https://github.com/apache/spark/commit/cf458279c4630ed0fb565b6836d10b909185309e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87463/
Test FAILed.

icexelloss committed:
Fix tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/918/
Test PASSed.

SparkQA commented:
**[Test build #87480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87480/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).

SparkQA commented:
**[Test build #87480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87480/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87480/
Test FAILed.

icexelloss commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/924/
Test PASSed.

SparkQA commented:
**[Test build #87490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87490/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).

SparkQA commented:
**[Test build #87490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87490/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87490/
Test PASSed.

icexelloss commented:
Resolved conflict and addressed @ueshin's comment. (Btw, I am fine with merging after Spark 2.3 RC passes, as that seems to be the priority now, just want to make sure this PR doesn't sit forever...)

HyukjinKwon commented:
Don't worry, I am keeping my eyes on this and I believe @ueshin too.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1118/
Test PASSed.

SparkQA commented:
**[Test build #87736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87736/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).

holdenk commented:
cc @BryanCutler

SparkQA commented:
**[Test build #87736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87736/testReport)** for PR 20295 at commit [`9ed3779`](https://github.com/apache/spark/commit/9ed3779b665c90e5bb25bc6636997a4b080c3d34).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87736/
Test PASSed.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasExec.scala:
```diff
@@ -75,28 +76,66 @@ case class FlatMapGroupsInPandasExec(
+      val index = dataAttributes.indexWhere(
+        childAttribute => attribute.semanticEquals(childAttribute))
+      dupGroupingIndices += index
+    }
```
nit:

```scala
val dupGroupingIndices = groupingAttributes.map { attribute =>
    dataAttributes.indexWhere(attribute.semanticEquals)
}
```

icexelloss replied:
Fixed. Thanks!

HyukjinKwon reviewed (commented):
Only nits I found while double checking.
LGTM too otherwise.

felixcheung commented on python/pyspark/sql/functions.py:
```diff
@@ -2253,6 +2253,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       ...     # key is a tuple of one numpy.int64, which is the value
+       ...     # of 'id' for the current group
+       ...     return pd.DataFrame([key + (pdf.v.mean(),)])
+       >>> df.groupby('id').apply(mean_udf).show()  #doctest: +SKIP
```
why skip all of these btw? why not run them so they can be tested?

HyukjinKwon replied:
I think it's because we couldn't find yet a min fix to enable the doctests only when PyArrow and Pandas are installed.

Maybe we can try to drop doctests right before we run `doctest.testmod` below conditionally but it's kind of a new approach to Spark as far as I know.

Will probably take a look for it separately soon.

felixcheung commented on python/pyspark/sql/types.py:
```diff
@@ -1725,6 +1737,29 @@ def _get_local_timezone():
+    from pyspark.sql.utils import require_minimum_pandas_version
+    require_minimum_pandas_version()
+
+    from pandas.api.types import is_datetime64tz_dtype
```
do we have tests for tese in tests.py?

icexelloss replied:
This function itself `_check_series_localize_timestamps` doesn't have a unit test, but it's called in various arrow/pandas_udf tests related to timestamps.

icexelloss commented:
@HyukjinKwon @felixcheung Thank you for the comments. I will take a look and hopefully address them today.

icexelloss commented on python/pyspark/worker.py:
```diff
@@ -149,18 +156,30 @@ def read_udfs(pickleSer, infile, eval_type):
+            udfs['f%d' % i] = udf
+            args = ["a[%d]" % o for o in arg_offsets]
+            call_udf.append("f%d(%s)" % (i, ", ".join(args)))
+            mapper_str = "lambda a: (%s)" % (", ".join(call_udf))
```
Aha good catch! I was surprised that this doesn't get caught in tests and found out the logic produced the same results unfortunately :(

icexelloss replied:
Fixed.

icexelloss committed:
Address PR comments

icexelloss commented on python/pyspark/sql/functions.py:
```diff
@@ -2253,6 +2253,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       be passed as the second argument. The grouping key will be passed as a tuple of numpy
+       data types, e.g., `numpy.int32` and `numpy.float64`. The data will still be passed in
+       as a `pandas.DataFrame` containing all columns from the original Spark DataFrame.
+       This is useful when the user doesn't want to hardcode grouping key in the function.
```
Changed to "does not"

icexelloss commented on python/pyspark/sql/functions.py:
```diff
@@ -2253,6 +2253,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       >>> df = spark.createDataFrame(
+       ...     [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
+       ...     ("id", "v"))  # doctest: +SKIP
+       >>> @pandas_udf("id long, v double", PandasUDFType.GROUP_MAP)  # doctest: +SKIP
```
Fixed.

icexelloss commented on python/pyspark/sql/functions.py:
```diff
@@ -2253,6 +2253,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       ... def mean_udf(key, pdf):
+       ...     # key is a tuple of one numpy.int64, which is the value
+       ...     # of 'id' for the current group
+       ...     return pd.DataFrame([key + (pdf.v.mean(),)])
```
Added

icexelloss commented on python/pyspark/worker.py:
```diff
@@ -149,18 +156,30 @@ def read_udfs(pickleSer, infile, eval_type):
+    if eval_type == PythonEvalType.SQL_GROUPED_MAP_PANDAS_UDF:
+        # Create function like this:
+        #   lambda a: f([a[0]], [a[0], a[1]])
+        assert num_udfs == 1
```
Added. Hopefully it's clear enough?

icexelloss commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasExec.scala:
```diff
@@ -75,28 +76,66 @@ case class FlatMapGroupsInPandasExec(
+    }
+
+    val extraGroupingSize = dupGroupingIndices.count(_ == -1)
+    (groupingAttributes zip dupGroupingIndices).foreach {
```
Fixed.

icexelloss commented:
Addressed all comments and manually tested the example in docstring.

SparkQA commented:
**[Test build #87968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87968/testReport)** for PR 20295 at commit [`722ed50`](https://github.com/apache/spark/commit/722ed508c280e63fb89cbdc891061c3c1928bf74).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1283/
Test PASSed.

SparkQA commented:
**[Test build #87968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87968/testReport)** for PR 20295 at commit [`722ed50`](https://github.com/apache/spark/commit/722ed508c280e63fb89cbdc891061c3c1928bf74).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87968/
Test PASSed.

HyukjinKwon commented:
Will merge this one if there's no more comments or not merged within few days.

BryanCutler commented on python/pyspark/worker.py:
```diff
@@ -91,10 +92,16 @@ def verify_result_length(*a):
-    def wrapped(*series):
+    def wrapped(key_series, value_series):
         import pandas as pd
+        argspec = inspect.getargspec(f)
```
should this also do `getfullargspec` for py3 like in udf.py?
maybe it would be useful to put a function in util.py, what do you guys think?

icexelloss replied:
Good point. Let me take a look at that.

ueshin commented:
@icexelloss Could you annotate `[SQL][PYTHON]` in the pr title please?

icexelloss committed:
refactor _get_argspec

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1328/
Test PASSed.

SparkQA commented:
**[Test build #88020 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88020/testReport)** for PR 20295 at commit [`c74ed05`](https://github.com/apache/spark/commit/c74ed05dec5c9b2521f5e37cf7e01e5ce873c779).

SparkQA commented:
**[Test build #88020 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88020/testReport)** for PR 20295 at commit [`c74ed05`](https://github.com/apache/spark/commit/c74ed05dec5c9b2521f5e37cf7e01e5ce873c779).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88020/
Test FAILed.

icexelloss commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1332/
Test PASSed.

SparkQA commented:
**[Test build #88025 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88025/testReport)** for PR 20295 at commit [`c74ed05`](https://github.com/apache/spark/commit/c74ed05dec5c9b2521f5e37cf7e01e5ce873c779).

BryanCutler commented on python/pyspark/sql/udf.py:
```diff
@@ -35,24 +37,29 @@ def _wrap_function(sc, func, returnType):
                                   sc.pythonVer, broadcast_vars, sc._javaAccumulator)
 
 
+def _get_argspec(f):
```
How about putting this in pyspark.util? It might be useful in places other than sql

icexelloss replied:
Make sense. Moved to pyspark.util.

SparkQA commented:
**[Test build #88025 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88025/testReport)** for PR 20295 at commit [`c74ed05`](https://github.com/apache/spark/commit/c74ed05dec5c9b2521f5e37cf7e01e5ce873c779).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88025/
Test PASSed.

icexelloss committed:
Move get_argspec to utils.py

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1354/
Test PASSed.

SparkQA commented:
**[Test build #88048 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88048/testReport)** for PR 20295 at commit [`59bdf20`](https://github.com/apache/spark/commit/59bdf200c6b922d18a64db0c7a67f9ec069c67ba).

icexelloss committed:
Move from pyspark.sql.utils to pyspark.util

SparkQA commented:
**[Test build #88049 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88049/testReport)** for PR 20295 at commit [`d51bc2e`](https://github.com/apache/spark/commit/d51bc2e9d0977177e0aff0aeab242376981b6b0d).

SparkQA commented:
**[Test build #88049 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88049/testReport)** for PR 20295 at commit [`d51bc2e`](https://github.com/apache/spark/commit/d51bc2e9d0977177e0aff0aeab242376981b6b0d).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88049/
Test FAILed.

icexelloss committed:
Fix style

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1355/
Test FAILed.

SparkQA commented:
**[Test build #88050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88050/testReport)** for PR 20295 at commit [`4b61f52`](https://github.com/apache/spark/commit/4b61f52497e9f356015ae7905fd08da785e90f08).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1356/
Test PASSed.

SparkQA commented:
**[Test build #88050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88050/testReport)** for PR 20295 at commit [`4b61f52`](https://github.com/apache/spark/commit/4b61f52497e9f356015ae7905fd08da785e90f08).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88050/
Test PASSed.

SparkQA commented:
**[Test build #88048 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88048/testReport)** for PR 20295 at commit [`59bdf20`](https://github.com/apache/spark/commit/59bdf200c6b922d18a64db0c7a67f9ec069c67ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88048/
Test PASSed.

HyukjinKwon commented:
Merged to master.

HyukjinKwon commented:
BTW, let's don't forget to fix the doc later ..

icexelloss commented:
Thanks all for review! @HyukjinKwon do you mean this doc?
https://github.com/apache/spark/blob/master/docs/sql-programming-guide.md#pyspark-usage-guide-for-pandas-with-apache-arrow

I can update it now or we can update later in batch before 2.4 release. What do you prefer?

HyukjinKwon commented:
Yup. Maybe we could do that when we are close to 2.4.

icexelloss commented:
Sounds good. Let's track in https://issues.apache.org/jira/browse/SPARK-23633

gatorsmile commented on python/pyspark/sql/tests.py:
```diff
@@ -4588,6 +4613,80 @@ def test_timestamp_dst(self):
+        self.assertPandasEqual(expected2, result2)
+
+        # Test complex groupby
+        result3 = df.groupby(df.id, df.v % 2).apply(udf2).sort('id', 'v').toPandas()
```
Any negative test case when the number of columns specified in groupby is different from the definition of udf (foo2)?

gatorsmile replied:
For end users, the misuse of this alternative functions could be common. For example, do we issue an appropriate error in the following cases?

- result3 = df.groupby(df.id).apply(udf2).sort('id', 'v').toPandas()
- result3 = df.groupby(df.id, df.v % 2, df.id).apply(udf2).sort('id', 'v').toPandas()
